### PR TITLE
Fix pipon definition merge

### LINKED ISSUE #412: Dictionary entries for lemmas with both noun and verb reading incorrectly conflated together
Following up on #190, so that inflected forms are now matched, the dictionary entries for the noun and verb reading are now conflated, which shouldn't take place, cf.

image

Unless this is so in the source, _winter_ (ni) and _it is winter_ (vii) should be two different dictionary entries also in itwêwina.

### PR DESCRIPTION
**Pull request by Madoshakalaka**
This PR fixes #412, where both the verb and the noun _pipon_ were wrongly associated with same 3 definitions. Some code didn't deal with things to enough resolution.

Now it's correct:

Screenshot from 2020-06-01 20-48-40

Potentially this also fixes some other cases.

I'll update production database after this is merged to master.

### CODE REVIEW

**Comment by Madoshakalaka** [comment]
@eddieantonio pretty technical changes. I do have comments if you want to scrutinize.

File: CreeDictionary/DatabaseManager/xml_importer.py [hunk 1]
```diff
@@ -462,22 +463,14 @@ def import_xmls(dir_name: Path, multi_processing: int = 1, verbose=True):
...
-        xml_lemmas = {t[0] for t in xml_lemma_pos_lcs}
-
-        xml_lemma_to_definition_src_dicts: Dict[str, List[Dict[str, Set[str]]]] = {
-            xml_lemma: [
-                xml_lemma_pos_lc_to_str_definitions[(xml_lemma,) + pos_lc_tuple]
-                for pos_lc_tuple in xml_lemma_to_pos_lc[xml_lemma]
-            ]
-            for xml_lemma in xml_lemmas
-        }
-
```
Comment (Madoshakalaka) [inline CreeDictionary/DatabaseManager/xml_importer.py:475]: A major task below is to find definitions from entries in the xml and associate them with machine generated wordforms.

`xml_lemma_to_definition_src_dicts` here is built and used in the process. The old way negalects pos/lc in the association process and uses the same definitions for homographs (the pipon case for e.g.).

Also this is mad comprehension to an extend of unneccessary. Thanks god it's finally deleted.

File: CreeDictionary/DatabaseManager/xml_importer.py [hunk 2]
```diff
@@ -462,22 +463,14 @@ def import_xmls(dir_name: Path, multi_processing: int = 1, verbose=True):
...
         # build wordforms and definition in db
         for generated_analysis, generated_wordforms in expanded[true_lemma_analysis]:
 
-            generated_lemma = fst_analysis_parser.extract_lemma(generated_analysis)
-            assert generated_lemma is not None
+            generated_lemma_lc = fst_analysis_parser.extract_lemma_and_category(
+                generated_analysis
+            )
+            assert generated_lemma_lc is not None
+            generated_lemma, generated_lc = generated_lemma_lc
```
Comment (Madoshakalaka) [inline CreeDictionary/DatabaseManager/xml_importer.py:473]: just to have an extra `generated_lc` to use below

File: CreeDictionary/DatabaseManager/xml_importer.py [hunk 3]
```diff
@@ -520,22 +513,43 @@ def import_xmls(dir_name: Path, multi_processing: int = 1, verbose=True):
...
+                        if does_lc_match_xml_entry(generated_lc, xml_pos, xml_lc):
+                            d_s_dicts.append(
+                                xml_lemma_pos_lc_to_str_definitions[
+                                    (generated_wordform, xml_pos, xml_lc)
+                                ]
+                            )
+                # The case when we don't have holographic entries in xml,
+                # The generated inflection doesn't have a definition
+                else:
+                    pass
```
Comment (Madoshakalaka) [inline CreeDictionary/DatabaseManager/xml_importer.py:538]: The new matching.

File: CreeDictionary/tests/DatabaseManager_tests/test_xml_importer_regressions.py [hunk 4]
```diff
@@ -23,17 +23,42 @@ def test_import_xml_common_analysis_definition_merge(shared_datadir):
...
+        .count()
+        == 1
+    )
+    assert (
+        Wordform.objects.filter(text="pipon", is_lemma=True, pos="V")
+        .get()
+        .definitions.all()
+        .count()
+        == 2
+    )
```
Comment (Madoshakalaka) [inline CreeDictionary/tests/DatabaseManager_tests/test_xml_importer_regressions.py:64]: unit/regression test

File: CreeDictionary/DatabaseManager/xml_importer.py [hunk 5]
```diff
@@ -520,22 +513,43 @@ def import_xmls(dir_name: Path, multi_processing: int = 1, verbose=True):
...
+                            text=str_definition,
+                            wordform=db_inflection,
+                        )
+                        assert definition_counter not in citations
+                        citations[definition_counter] = set(source_strings)
 
-                            definition_counter += 1
-                            db_definitions.append(db_definition)
+                        definition_counter += 1
+                        db_definitions.append(db_definition)
```
Comment (Madoshakalaka) [inline CreeDictionary/DatabaseManager/xml_importer.py:552]: black formatting

File: CreeDictionary/DatabaseManager/xml_importer.py (hunk 1 above)
Comment (eddieantonio) [reply CreeDictionary/DatabaseManager/xml_importer.py:475]: Deleting code is always great!

File: CreeDictionary/DatabaseManager/xml_importer.py [hunk 6]
```diff
@@ -520,22 +513,43 @@ def import_xmls(dir_name: Path, multi_processing: int = 1, verbose=True):
                         keyword_counter += 1
 
                 # now we create definition for all (possibly inflected) entries in xml that are forms of this lemma.
-                d_s_dicts = xml_lemma_to_definition_src_dicts.get(generated_wordform)
+                # d_s_dicts = xml_lemma_to_definition_src_dicts.get(generated_wordform)
```
Comment (eddieantonio) [inline CreeDictionary/DatabaseManager/xml_importer.py:516]:
```suggestion
```

File: CreeDictionary/DatabaseManager/xml_importer.py [hunk 7]
```diff
@@ -520,22 +513,43 @@ def import_xmls(dir_name: Path, multi_processing: int = 1, verbose=True):
...
+
+                # try to match our generated wordform to entries in the xml file,
+                # in order to get its definition from the entries
+                d_s_dicts: List[dict] = []
+
+                # first get homographic entries from the xml file
+                pos_lc_tuples_in_xml = xml_lemma_to_pos_lc.get(generated_wordform)
+
+                # The case when we do have holographic entries in xml,
```
Comment (eddieantonio) [inline CreeDictionary/DatabaseManager/xml_importer.py:525]:
```suggestion
                # The case when we do have homographic entries in xml,
```

well holo everyone

File: CreeDictionary/DatabaseManager/xml_importer.py (hunk 3 above)
Comment (eddieantonio) [inline CreeDictionary/DatabaseManager/xml_importer.py:538]:
```suggestion
```

File: CreeDictionary/DatabaseManager/xml_importer.py (hunk 6 above)
Comment (Madoshakalaka) [reply CreeDictionary/DatabaseManager/xml_importer.py:516]: @eddieantonio LOOLOOLOLL I did notice that line of undeleted commented out code when I was reviewing. And I was thinking "fine I'll just leave it there Eddie won't notice"

**Comment by aarppe** [comment]
Yes, it now works as it should for _kîsikâw_ as well, see:

image

This also seems to mostly be the case for also _kimiwan_,

image

... though there I think the major cause of the remaining mix-up is in the XML source.

N.B. This should **NOT** be fixed programmatically, but in the source (XML or otherwise), by linguists. The reason for this is probably that the comparison script I used didn't properly disambiguate word-class-wise ambiguous dictionary entries.

**Comment by aarppe** [comment]
Generally, there is a small but non-insignificant number of dictionary entries which are ambiguous for the head word (e.g. II verbs and nouns, some II and TA verbs). Because of this, any combination of the dictionary entries may need to be sensitive as deep down as the inflectional subcategory (sometimes the general word class is sufficient, but not always). In the XML, the inflectional subcategory is provided in the `lc` variable. Something similar(ly named) will be used in whatever format the new, aggregated crk dictionary source will look like.

**Comment by Madoshakalaka** [comment]
@aarppe get it

**Pull request merged**